sage.py

```diff
@@ -481,1274 +423,920 @@ def getepisodes(action_seq, mcat, plot, debug=False):
-                        alerts.append(episodes)
-                        #pass
-                    else:
-                        rest = (-1,-1)
-                       
-                        for i in range(len(cuts)):
-                            start, end = 0, 0
-                            if i == 0:
-                                start = 0
-                                end = cuts[i]  
-                            else:
-                                start = cuts[i-1]+1
-                                end = cuts[i]
-                            rest = (end+1, len(ep)-1)
-                            al = episodes[start:end+1]
-                            if len(al) < 2:
-                                #print('discarding symbol ', [x[2] for x in al])#, start, end, len(episodes))
-                                continue
-                            att = atta+'->'+victim+'-'+str(c)
-                            #print('---', att, [x[2] for x in al])
-                            keys.append(att)
-                            alerts.append(al)
-                            c+=1
-                        #print('--', ep[rest[0]: rest[1]+1])
-                        al = episodes[rest[0]: rest[1]+1]
-                        if len(al) < 2:
-                            #print('discarding symbol ', [x[2] for x in al]) # TODO This one is not cool1
-                            continue
-                        att = atta+'->'+victim+'-'+str(c)
-                        #print('---', att, [x[2] for x in al])
-                        keys.append(att)
-                        alerts.append(al)
-    print('\n# sub-sequences', len(keys))
-    return (alerts, keys)
-
-## 27 Aug 2020: Generating traces for flexfringe 
-def generate_traces(alerts, keys, datafile, test_ratio=0.0):
-    victims = alerts
-    al_services = [[most_frequent(y[6]) for y in x] for x in victims]
-    print('----- All unique servcies -----')
-    print(set([item for sublist in al_services for item in sublist]))
+            # Cut episode sequence when a low-severity episode follows a high-severity episode
+            count = 0
+            mcats = [epi[2] for epi in episodes]
+            cuts = [i for i in range(len(episodes) - 1) if (len(str(mcats[i])) > len(str(mcats[i + 1])))]  # (ep[i] > 100 and ep[i+1] < 10)]
+
+            rest = (0, len(episodes) - 1)
+            for j in range(len(cuts)):
+                start = 0 if j == 0 else cuts[j - 1] + 1
+                end = cuts[j]
+                rest = (end + 1, len(episodes) - 1)
+                subsequence = episodes[start:end+1]
+                if len(subsequence) < 2:
+                    continue
+                _subsequences[attacker_victim + '-' + str(count)] = subsequence
+                count += 1
+            subsequence = episodes[rest[0]:rest[1]+1]
+            if len(subsequence) < 2:
+                # print('discarding symbol ', [x[2] for x in al]) # TODO This one is not cool1
```

Comment:
Is the TODO still relevant?

Reply:
Yep. There is a problem in this code: when splitting, (if I remember correctly) if a sequence goes like [low low medium high low], then [low low medium high] is saved but the last [low] is just discarded. We probably shouldnt lose alerts like this. But idk what to do with a single event either. Maybe we keep them regardless?

Reply:
Hmm, for medium and high - maybe, for low - not sure. But for this one episode here, I don't think it is really useful. If AGs accentuate high-severity objectives, then this `high` in your sequence would be the end of an attack attempt. The next one does not give that much information. Also, wouldn't it mean giving a trace to FlexFringe with one symbol? I am not sure it will be happy to receive such one symbol, since it doesn't even have a transition